File: circle-packing/js_circle-packing/circle-packing.js
var canvas = document.querySelector('canvas');
var context = canvas.getContext('2d');
var mirror = document.getElementById('mirror');

var size = 800;
var dpr = window.devicePixelRatio;
canvas.width = mirror.width = size * dpr;
canvas.height = mirror.width = size * dpr;
context.scale(dpr, dpr);

context.lineWidth = 2;

var circles = [];
var minRadius = 3;
var maxRadius = size / 5;
var totalCircles = Math.floor(size * 1.35);
var createCircleAttempts = 500;


function createAndDrawCircle() {

    var newCircle;
    var circleSafeToDraw = false;

    for (var tries = 0; tries < createCircleAttempts; tries++) {
        var newCircle = {
            x: Math.floor(Math.random() * size),
            y: Math.floor(Math.random() * size),
            radius: minRadius
        };

        if (doesCircleHaveACollision(newCircle)) {
            continue;
        } else {
            circleSafeToDraw = true;
            break;
        }

        if (!circleSafeToDraw) {
            return;
        }
    }

    for (var radiusSize = minRadius; radiusSize < maxRadius; radiusSize++) {
        newCircle.radius = radiusSize;

        if (doesCircleHaveACollision(newCircle)) {
            newCircle.radius--;
            break;
        }
    }
    
    circles.push(newCircle);
    context.beginPath();
    context.arc(newCircle.x, newCircle.y, newCircle.radius, 0, 2 * Math.PI);
    context.stroke();

}

function doesCircleHaveACollision(circle) {

    for (var i = 0; i< circles.length; i++) {
        var otherCircle = circles[i];
        var a = circle.radius + otherCircle.radius;
        var x = circle.x - otherCircle.x;
        var y = circle.y - otherCircle.y;

        if (a >= Math.sqrt((x * x) + (y * y))) {
            return true;
        }

        if (circle.x + circle.radius >= size ||
            circle.x - circle.radius <= 0) {
            
            return true;
        }
    

        if (circle.y + circle.radius >= size ||
            circle.y - circle.radius <= 0) {
            
            return true;
        }
    }

    return false;
}

for (var i = 0; i < totalCircles; i++) {
    createAndDrawCircle();
}

mirror.addEventListener('contextmenu', function (e) {
    var dataURL = canvas.toDataURL('image/png');
    mirror.src = dataURL;
});

File: circle-packing/py_circle-packing/circle-packing.py
from math import floor, sqrt
from random import random

import pygame

SIZE = 800
MIN_RADIUS = 3
MAX_RADIUS = SIZE // 3
TOTAL_CIRCLES = floor(SIZE * 1.1)
CREATE_CIRCLE_ATTEMPTS = 500
LINE_WIDTH = 2

NON_INTERACTIVE_UPDATE_EVENT = pygame.event.Event(pygame.USEREVENT, attr1='NON_INTERACTIVE_UPDATE_EVENT')

pygame.init()
screen = pygame.display.set_mode((SIZE, SIZE))
clock = pygame.time.Clock()
done = False

class Circle:
    def __init__(self, x, y, radius):
        self.x = x
        self.y = y
        self.radius = radius

circles = []

def does_circle_have_a_collision(circle):
    for i in range(0, len(circles)):
        other_circle = circles[i]
        a = circle.radius + other_circle.radius
        x = circle.x - other_circle.x
        y = circle.y - other_circle.y

        if a >= sqrt((x * x) + (y * y)):
            return True

        if circle.x + circle.radius >= SIZE or circle.x - circle.radius <= 0:
            return True

        if circle.y + circle.radius >= SIZE or circle.y - circle.radius <= 0:
            return True

    return False

def create_circle():

    new_circle = None

    for _ in range(0, CREATE_CIRCLE_ATTEMPTS):

        new_circle = Circle(floor(random() * SIZE), floor(random() * SIZE), MIN_RADIUS)

        if does_circle_have_a_collision(new_circle):
            continue
        else:
            break

    for radius_size in range(MIN_RADIUS, MAX_RADIUS):
        new_circle.radius = radius_size

        if does_circle_have_a_collision(new_circle):
            new_circle.radius -= 1
            break

    circles.append(new_circle)

def draw_circles():
    for circle in circles:
        pygame.draw.circle(screen, (0, 0, 0, 0), (circle.x, circle.y), circle.radius, LINE_WIDTH)


pygame.event.post(NON_INTERACTIVE_UPDATE_EVENT)
screen.fill((255, 255, 255))

while not done:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            done = True

        if event == NON_INTERACTIVE_UPDATE_EVENT:
            if len(circles) < TOTAL_CIRCLES:
                create_circle()
                draw_circles()
            pygame.display.flip()
            clock.tick(480)
            pygame.event.post(NON_INTERACTIVE_UPDATE_EVENT)
   
